consensus/src/pipeline/virtual_processor/processor.rs

```diff
@@ -480,10 +493,53 @@ impl VirtualStateProcessor {
         (virtual_parents, ghostdag_data)
     }
 
-    pub fn build_block_template(self: &Arc<Self>, miner_data: MinerData, mut txs: Vec<Transaction>) -> BlockTemplate {
+    pub fn validate_mempool_transaction_and_populate(&self, mutable_tx: &mut MutableTransaction) -> TxResult<()> {
+        let virtual_read = self.virtual_stores.read();
+        let virtual_state = virtual_read.state.get().unwrap();
+        let virtual_utxo_view = &virtual_read.utxo_set;
+        let virtual_daa_score = virtual_state.daa_score;
+        let virtual_past_median_time = virtual_state.past_median_time;
+
+        self.transaction_validator.validate_tx_in_isolation(&mutable_tx.tx)?;
```

Comment:
The first part can be done outside the lock. Anyway, this part requires more design.